allennlp/training/trainer.py

```diff
@@ -145,136 +161,243 @@ def train(self) -> None:
                 if parameter.requires_grad:
                     parameter.register_hook(clip_function)
 
-        logger.info("Beginning training.")
+    def _rescale_gradients(self) -> None:
+        """
+        Performs gradient rescaling. Is a no-op if gradient rescaling is not enabled.
+        """
+        if self._grad_norm:
+            clip_grad_norm(self._model.parameters(), self._grad_norm)
+
+    def _compute_loss(self, batch: torch.Tensor, for_training: bool) -> torch.Tensor:
+        """
+        Does a forward pass on the given batch and returns the ``loss`` value in the result.
+        Also performs backpropagation if ``for_training`` is ``True``.
+        """
+        output_dict = self._forward(batch, for_training=for_training)
+
+        try:
+            loss = output_dict["loss"]
+            # TODO(joelgrus): add in regularization
+        except KeyError:
+            raise ConfigurationError("The model you are trying to optimize does not contain a"
+                                     " 'loss' key in the output of model.forward(inputs).")
+
+        # Backpropagate the loss if we're training.
+        if for_training:
+            loss.backward()
+
+        return loss
+
+    def _get_metrics_for_batch(self, total_loss: float, batch_num: int, reset: bool = False) -> dict:
+        """
+        Gets the metrics but sets ``"loss"`` to
+        the total loss divided by the ``batch_num`` so that
+        the ``"loss"`` metric is "average loss per batch".
+        """
+        metrics = self._model.get_metrics(reset)
+        metrics["loss"] = float(total_loss / batch_num)
+        return metrics
+
+    def _train_epoch(self, epoch: int) -> dict:
+        """
+        Trains one epoch and returns metrics.
+        """
+        logger.info("Epoch %d/%d", epoch, self._num_epochs - 1)
+        train_loss = 0.0
+        # Set the model to "train" mode.
+        self._model.train()
+
+        # Get tqdm for the training batches
+        train_generator = self._iterator(self._train_dataset, num_epochs=1)
         num_training_batches = self._iterator.get_num_batches(self._train_dataset)
+        train_generator_tqdm = tqdm.tqdm(train_generator,
+                                         disable=self._no_tqdm,
+                                         total=num_training_batches)
+        self._last_log = time.time()
+        batch_num = 0
+
+        logger.info("Training")
+        for batch in train_generator_tqdm:
+            # Increment batch and zero gradient
+            batch_num += 1
+            self._optimizer.zero_grad()
+
+            # Compute loss for batch and backpropagate
+            loss = self._compute_loss(batch, for_training=True)
+
+            # Make sure Variable is on the cpu before converting to numpy.
+            # .cpu() is a no-op if you aren't using GPUs.
+            train_loss += loss.data.cpu().numpy()
+
+            # Rescale gradients
```

Comment:
Most of these comments are just annotating a single line now, where the function being called already has a descriptive name.  I think you can remove these comments.

Reply:
done